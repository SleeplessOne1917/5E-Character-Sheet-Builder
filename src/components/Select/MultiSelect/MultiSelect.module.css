.list {
    list-style: none;
    background-color: var(--heaven-light);
    margin-block-start: 0;
    margin-block-end: 0;
    padding: 0;
    border: 1px solid var(--davy);
    font-size: inherit;
    border-radius: 0.75rem;
    overflow-y: auto;
    overflow-x: hidden;
    max-height: 15rem;
    width: fit-content;
    color: var(--davy);
}

.list-box-container {
    width: fit-content;
}

.list-box-container.error {
    border: 3px solid var(--error-red);
    border-radius: 1rem;
    padding: 0.1rem;
}

.item {
    padding: 0.25rem 0.5rem;
    cursor: pointer;
    text-align: center;
}

.item:first-of-type {
    border-top-left-radius: 0.75rem;
    border-top-right-radius: 0.75rem;
}

.item:last-of-type {
    border-bottom-left-radius: 0.75rem;
    border-bottom-right-radius: 0.75rem;
}

.item:focus-visible {
    outline: 0;
    background-color: var(--highlight-light);
}

.selected {
    background-color: var(--highlight)  !important;
}

.selected:focus-visible {
    background-color: var(--highlight-hover) !important;
}

@media (hover: hover) {
    .item:hover {
        background-color: var(--highlight-light);
    }

    .selected:hover {
        background-color: var(--highlight-hover) !important;
    }
}