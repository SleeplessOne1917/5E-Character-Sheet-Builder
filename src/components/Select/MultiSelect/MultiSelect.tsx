import classes from './MultiSelect.module.css';
import selectClasses from '../Select.module.css';
import { useCallback, useRef, useState, KeyboardEvent } from 'react';
import Option from '../Option';
import { handleKeyDownEvent } from '../../../services/handlerService';

type MultiSelectProps = {
	id?: string;
	label?: string;
	labelFontSize?: string;
	touched?: boolean;
	error?: string;
	fontSize?: string;
	options: Option[];
	values: (string | number)[];
	onSelect: (values: (string | number)[]) => void;
};

const MultiSelect = ({
	id,
	label,
	labelFontSize = '1.5rem',
	touched,
	error,
	fontSize,
	options,
	values,
	onSelect
}: MultiSelectProps) => {
	const [selectedValues, setSelectedValues] = useState(values ?? []);
	const [focusIndex, setFocusIndex] = useState(0);

	const listItemRefs = useRef<HTMLLIElement[]>([]);

	const handleListKeyDown = useCallback(
		(e: KeyboardEvent) => {
			switch (e.code) {
				case 'ArrowUp':
					e.preventDefault();
					if (focusIndex > 0) {
						setFocusIndex(prev => {
							const returnVal = prev - 1;
							listItemRefs.current[returnVal].focus();
							return returnVal;
						});
					}
					break;
				case 'ArrowDown':
					e.preventDefault();
					if (focusIndex < listItemRefs.current.length - 1) {
						setFocusIndex(prev => {
							const returnVal = prev + 1;
							listItemRefs.current[returnVal].focus();
							return returnVal;
						});
					}
					break;
				default:
					break;
			}
		},
		[focusIndex]
	);
	const handleChange = useCallback(
		(changeValue: number | string) => {
			setSelectedValues(prev => {
				const returnValues = prev.includes(changeValue)
					? prev.filter(val => val !== changeValue)
					: [...prev, changeValue];

				onSelect(returnValues);

				return returnValues;
			});
		},
		[onSelect, setSelectedValues]
	);

	const handleItemKeyDown = useCallback(
		(event: KeyboardEvent<HTMLLIElement>, value: number | string) => {
			handleKeyDownEvent(event, () => handleChange(value));
		},
		[handleChange]
	);

	const handleFocus = useCallback((index: number) => {
		setFocusIndex(index);
	}, []);

	return (
		<div
			data-testid={'multi-select'}
			className={selectClasses['list-box-and-error-container']}
		>
			{label && id && (
				<label id={`${id}-label`} style={{ fontSize: labelFontSize }}>
					{label}
				</label>
			)}
			<div
				className={`${classes['list-box-container']}${
					touched && error ? ` ${classes.error}` : ''
				}`}
				style={{ fontSize: fontSize ? fontSize : '1.2rem' }}
			>
				<ul
					tabIndex={-1}
					role="listbox"
					aria-labelledby={`${id}-label`}
					aria-multiselectable="true"
					className={classes.list}
					onKeyDown={handleListKeyDown}
					data-testid="multi-select-list"
				>
					{options.map((option, index) => (
						<li
							key={option.value}
							tabIndex={0}
							role="option"
							aria-selected={selectedValues.includes(option.value)}
							className={`${classes.item}${
								selectedValues.includes(option.value)
									? ` ${classes.selected}`
									: ''
							}`}
							onClick={() => handleChange(option.value)}
							onKeyDown={event => handleItemKeyDown(event, option.value)}
							ref={element => {
								listItemRefs.current[index] = element as HTMLLIElement;
							}}
							data-testid={`data-${option.value}`}
							onFocus={() => handleFocus(index)}
						>
							{option.label}
						</li>
					))}
				</ul>
			</div>
			{touched && error && (
				<div className={selectClasses['error-message']}>{error}</div>
			)}
		</div>
	);
};

export default MultiSelect;
