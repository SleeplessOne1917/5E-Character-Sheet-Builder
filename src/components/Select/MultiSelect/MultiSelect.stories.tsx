import { ComponentMeta, ComponentStory } from '@storybook/react';
import MultiSelect from './MultiSelect';

export default {
	title: 'Components/Select/MultiSelect',
	component: MultiSelect,
	args: {
		id: 'stuff',
		label: 'Stuff',
		options: [
			{ value: 'foo', label: 'Foo' },
			{ value: 'bar', label: 'Bar' },
			{ value: 'baz', label: 'Baz' }
		]
	}
} as ComponentMeta<typeof MultiSelect>;

const Template: ComponentStory<typeof MultiSelect> = args => (
	<MultiSelect {...args} />
);

export const Default = Template.bind({});

export const LotsOfOptions = Template.bind({});

LotsOfOptions.args = {
	options: [
		{ value: 'foo', label: 'Foo' },
		{ value: 'bar', label: 'Bar' },
		{ value: 'baz', label: 'Baz' },
		{ value: 'spam', label: 'Eggs' },
		{ value: 'barbarian', label: 'Barbarian' },
		{ value: 'druid', label: 'Druid' },
		{ value: 'bard', label: 'Bard' },
		{ value: 'fighter', label: 'Fighter' },
		{ value: 'wizard', label: 'Wizard' }
	]
};

export const Error = Template.bind({});

Error.args = {
	touched: true,
	error: 'Your values suck'
};
