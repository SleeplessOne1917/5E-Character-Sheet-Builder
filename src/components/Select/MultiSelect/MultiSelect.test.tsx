import { render, screen } from '@testing-library/react';
import * as stories from './MultiSelect.stories';
import { composeStories } from '@storybook/testing-react';

const { Default, Error, LotsOfOptions } = composeStories(stories);

describe('renders correctly', () => {
	it('by default', () => {
		render(<Default />);

		expect(screen.getByTestId('multi-select')).toMatchSnapshot();
	});

	it('with a lot of options', () => {
		render(<LotsOfOptions />);

		expect(screen.getByTestId('multi-select')).toMatchSnapshot();
	});

	it('with error', () => {
		render(<Error />);

		expect(screen.getByTestId('multi-select')).toMatchSnapshot();
	});
});
