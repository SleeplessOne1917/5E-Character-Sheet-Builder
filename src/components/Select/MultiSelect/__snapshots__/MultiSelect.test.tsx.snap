// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<div
  class="list-box-and-error-container"
  data-testid="multi-select"
>
  <label
    id="stuff-label"
    style="font-size: 1.5rem;"
  >
    Stuff
  </label>
  <div
    class="list-box-container"
    style="font-size: 1.2rem;"
  >
    <ul
      aria-labelledby="stuff-label"
      aria-multiselectable="true"
      class="list"
      data-testid="multi-select-list"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="item"
        data-testid="data-foo"
        role="option"
        tabindex="0"
      >
        Foo
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-bar"
        role="option"
        tabindex="0"
      >
        Bar
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-baz"
        role="option"
        tabindex="0"
      >
        Baz
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders correctly with a lot of options 1`] = `
<div
  class="list-box-and-error-container"
  data-testid="multi-select"
>
  <label
    id="stuff-label"
    style="font-size: 1.5rem;"
  >
    Stuff
  </label>
  <div
    class="list-box-container"
    style="font-size: 1.2rem;"
  >
    <ul
      aria-labelledby="stuff-label"
      aria-multiselectable="true"
      class="list"
      data-testid="multi-select-list"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="item"
        data-testid="data-foo"
        role="option"
        tabindex="0"
      >
        Foo
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-bar"
        role="option"
        tabindex="0"
      >
        Bar
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-baz"
        role="option"
        tabindex="0"
      >
        Baz
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-spam"
        role="option"
        tabindex="0"
      >
        Eggs
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-barbarian"
        role="option"
        tabindex="0"
      >
        Barbarian
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-druid"
        role="option"
        tabindex="0"
      >
        Druid
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-bard"
        role="option"
        tabindex="0"
      >
        Bard
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-fighter"
        role="option"
        tabindex="0"
      >
        Fighter
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-wizard"
        role="option"
        tabindex="0"
      >
        Wizard
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders correctly with error 1`] = `
<div
  class="list-box-and-error-container"
  data-testid="multi-select"
>
  <label
    id="stuff-label"
    style="font-size: 1.5rem;"
  >
    Stuff
  </label>
  <div
    class="list-box-container error"
    style="font-size: 1.2rem;"
  >
    <ul
      aria-labelledby="stuff-label"
      aria-multiselectable="true"
      class="list"
      data-testid="multi-select-list"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="item"
        data-testid="data-foo"
        role="option"
        tabindex="0"
      >
        Foo
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-bar"
        role="option"
        tabindex="0"
      >
        Bar
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-baz"
        role="option"
        tabindex="0"
      >
        Baz
      </li>
    </ul>
  </div>
  <div
    class="error-message"
  >
    Your values suck
  </div>
</div>
`;
